src/cmap/auth/mongodb_oidc/callback_workflow.ts

```diff
@@ -0,0 +1,227 @@
+      // No entry means to start with the step one saslStart.
+      const result = await connection.commandAsync(
+        ns(credentials.source),
+        startCommandDocument(credentials),
+        undefined
+      );
+      const stepOne = BSON.deserialize(result.payload.buffer) as OIDCMechanismServerStep1;
+      // Call the request callback and finish auth.
+      return this.requestAndFinish(connection, credentials, stepOne, result.conversationId);
+    }
+  }
+
+  /**
+   * Execute the refresh callback if it exists, otherwise the request callback, then
+   * finish the authentication.
+   */
+  private async refreshAndFinish(
+    connection: Connection,
+    credentials: MongoCredentials,
+    stepOneResult: OIDCMechanismServerStep1,
+    tokenResult: OIDCRequestTokenResult,
+    conversationId?: number
+  ): Promise<Document> {
+    const refresh = credentials.mechanismProperties.REFRESH_TOKEN_CALLBACK;
+    // If a refresh callback exists, use it. Otherwise use the request callback.
+    if (refresh) {
+      const result: OIDCRequestTokenResult = await refresh(
+        credentials.username,
+        stepOneResult,
+        tokenResult,
+        TIMEOUT_MS
+      );
+      // Validate the result.
+      if (!result || !result.accessToken) {
+        throw new MongoMissingCredentialsError(
+          'REFRESH_TOKEN_CALLBACK must return a valid object with an accessToken'
+        );
+      }
+      // Cache a new entry and continue with the saslContinue.
+      this.cache.addEntry(connection.address, credentials.username || '', result, stepOneResult);
+      return finishAuth(result, conversationId, connection, credentials);
+    } else {
+      // Fallback to using the request callback.
+      return this.requestAndFinish(connection, credentials, stepOneResult, conversationId);
+    }
+  }
+
+  /**
+   * Execute the request callback and finish authentication.
+   */
+  private async requestAndFinish(
+    connection: Connection,
+    credentials: MongoCredentials,
+    stepOneResult: OIDCMechanismServerStep1,
+    conversationId?: number
+  ): Promise<Document> {
+    // Call the request callback.
+    const request = credentials.mechanismProperties.REQUEST_TOKEN_CALLBACK;
+    // Always clear expired entries from the cache on each finish as cleanup.
+    this.cache.deleteExpiredEntries();
+    if (request) {
+      const tokenResult = await request(credentials.username, stepOneResult, TIMEOUT_MS);
+      // Validate the result.
+      if (!tokenResult || !tokenResult.accessToken) {
+        throw new MongoMissingCredentialsError(
+          'REQUEST_TOKEN_CALLBACK must return a valid object with an accessToken'
+        );
+      }
+      // Cache a new entry and continue with the saslContinue.
+      this.cache.addEntry(
+        connection.address,
+        credentials.username || '',
+        tokenResult,
+        stepOneResult
+      );
+      return finishAuth(tokenResult, conversationId, connection, credentials);
+    } else {
```

Comment:
can we invert this? `if (!request) { throw }` then un-nest what's current L149-163?

Reply:
Inverted.